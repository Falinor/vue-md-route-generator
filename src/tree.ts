export type Tree<T> = { value: T; children?: Tree<T>[] }

type ForEachFunction<T> = (node: T) => void

export function forEach<T = any>(tree: Tree<T>, fn: ForEachFunction<T>): void {
  const { value, children } = tree
  fn(value)

  if (!children) {
    return
  }

  children.forEach(child => {
    forEach(child, fn)
  })
}

type MapFunction<T, U> = (value: T, children?: Tree<T>[]) => U

export function map<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): Tree<U> {
  const { value, children } = tree

  if (!children) {
    return { value: fn(value) }
  }

  return {
    value: fn(value, children),
    children: children.map(child => map(child, fn))
  }
}

export function flatten<T>(tree: Tree<T>): T[] {
  function flattenRec<T>(acc: T[], node: Tree<T>): T[] {
    acc.push(node.value)
    if (node.children) {
      acc.concat(...node.children.map(child => flattenRec(acc, child)))
    }
    return acc
  }
  return flattenRec<T>([], tree)
}

export function flatMap<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): U[] {
  return flatten(map(tree, fn))
}

export type SortFunction<T> = (a: Tree<T>, b: Tree<T>) => number

export function sort<T = any>(tree: Tree<T>, fn: SortFunction<T>): Tree<T> {
  const { value, children } = tree

  if (!children) {
    return tree
  }

  return {
    value,
    children: [...children].map(child => sort(child, fn)).sort(fn)
  }
}

export type FilterFunction<T> = (value: T, children?: Tree<T>[]) => boolean

export function filter<T = any>(
  tree: Tree<T>,
  fn: FilterFunction<T>
): Tree<T> | null {
  if (!fn(tree.value)) {
    return null
  }

  if (!tree.children) {
    return tree
  }

  return {
    value: tree.value,
    children: tree.children
      .map(child => filter(child, fn))
      .filter(v => v !== null) as Tree<T>[]
  }
}
