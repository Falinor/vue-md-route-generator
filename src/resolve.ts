import parseFrontMatter from 'gray-matter'
import isEmpty from 'lodash.isempty'
import path from 'path'

import { filter, map, sort, Tree } from './tree'

export interface PageMeta {
  name: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string | null
  routeMeta?: unknown
}

export interface File {
  path: string
  type: 'directory' | 'file'
}

export type FileTree = Tree<File>
export type PageMetaTree = Tree<PageMeta>

export function resolveRoutePaths(
  tree: FileTree,
  importPrefix: string,
  readFile: (path: string) => string
): PageMetaTree {
  return sortByMeta(
    filterDirectoryMeta(
      map(tree, (file: File, children?: FileTree[]) => {
        const segments: string[] = file.path.split('/')
        const meta: PageMeta = {
          name: pathToName(segments),
          path: pathToRoute(segments, 0, false),
          pathSegments: toActualPath(segments),
          component: null,
          specifier: pathToSpecifier(segments)
        }

        if (file.type === 'file') {
          const content = readFile(file.path)
          const { data } = parseFrontMatter(content)
          if (!isEmpty(data)) {
            meta.routeMeta = data
          }
          meta.component = importPrefix + file.path
        }

        if (file.type === 'directory' && children?.length) {
          const metaFile = children.find(
            child =>
              child.value.type === 'file' &&
              basename(child.value.path) === '__meta__'
          )
          if (metaFile !== undefined) {
            const content = readFile(metaFile.value.path)
            const { data } = parseFrontMatter(content)
            if (!isEmpty(data)) {
              meta.routeMeta = data
            }
          }
        }

        return meta
      })
    )
  )
}

function sortByMeta(tree: PageMetaTree): PageMetaTree {
  return sort(tree, (a: PageMetaTree, b: PageMetaTree): number => {
    const routeMeta1 = a.value.routeMeta
    const routeMeta2 = b.value.routeMeta

    if (hasOrder(routeMeta1)) {
      if (hasOrder(routeMeta2)) {
        return routeMeta1.order - routeMeta2.order
      }
    }

    if (hasOrder(routeMeta2)) {
      return 1
    }

    return 0
  })
}

function hasOrder(obj: unknown): obj is { order: number } {
  return typeof obj === 'object' && obj !== null && obj.hasOwnProperty('order')
}

function filterDirectoryMeta(tree: PageMetaTree): PageMetaTree {
  return filter(
    tree,
    meta => !meta.component?.endsWith('__meta__.md')
  ) as PageMetaTree
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.md` from the last path
 * - Omit if the last segment is `index`
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])

  const segmentsWithoutLast = segments.slice(0, -1)
  return isOmittable(last)
    ? segmentsWithoutLast
    : segmentsWithoutLast.concat(last)
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  return segments
    .slice(0, -1)
    .concat(basename(last))
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const name = pathToName(segments)
  const replaced = name
    .replace(/(^|[^a-zA-Z])([a-zA-Z])/g, (_, a, b) => {
      return a + b.toUpperCase()
    })
    .replace(/[^a-zA-Z0-9]/g, '')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return (
    prefix +
    toActualPath(segments)
      .slice(parentDepth)
      .join('/')
  )
}

function basename(filename: string): string {
  return path.basename(filename, '.md')
}
