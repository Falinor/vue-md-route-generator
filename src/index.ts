import directoryTree, { DirectoryTree } from 'directory-tree'
import fs from 'fs'
import { basename, dirname, isAbsolute, join } from 'path'
import prettier from 'prettier'

import { FileTree, resolveRoutePaths } from './resolve'
import { createRoutes, RouteString } from './template/routes'

export interface GenerateConfig {
  folders: string[]
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
}

function relativeTo(prefix: string, path: string): string {
  const segments = path.split('/')
  const index = segments.findIndex(seg => seg === prefix)
  return segments.slice(index).join('/')
}

function toPathTree(directoryTree: DirectoryTree, prefix: string): FileTree {
  return {
    value: {
      path: isAbsolute(directoryTree.path)
        ? relativeTo(prefix, directoryTree.path)
        : directoryTree.path,
      type: directoryTree.type
    },
    children: directoryTree?.children?.map(child => toPathTree(child, prefix))
  }
}

export function generateRoutes({
  folders,
  importPrefix = '@/assets/',
  dynamicImport = true,
  chunkNamePrefix = ''
}: GenerateConfig): string {
  const routeStrings: RouteString[] = folders.map(folder => {
    const dirTree = directoryTree(folder, {
      extensions: /\.md$/,
      normalizePath: true
    })
    const pathTree = toPathTree(dirTree, basename(folder))

    const meta = resolveRoutePaths(pathTree, importPrefix, file => {
      return fs.readFileSync(join(dirname(folder), file), 'utf8')
    })
    return createRoutes(meta, dynamicImport, chunkNamePrefix)
  })
  const rendererImport: string = `
  import Vue from 'vue'
  
  Vue.component('renderer', resolve => {
    setTimeout(() => resolve({
      render: createElement => createElement('router-view')
    }))
  })
  `
  const imports: string = [rendererImport]
    .concat(routeStrings.map(rs => rs.imports))
    .join('\n')
  const routes: string = routeStrings.map(rs => rs.code).join(',\n')
  return prettier.format(`${imports}\n\nexport default [${routes}]`, {
    parser: 'babel',
    semi: false,
    singleQuote: true
  })
}
