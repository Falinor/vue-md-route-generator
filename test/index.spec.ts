import path from 'path'

import { generateRoutes } from '../src'

it('Integration test', () => {
  const res = generateRoutes({
    folders: [
      path.resolve(__dirname, 'fixtures', 'foo'),
      path.resolve(__dirname, 'fixtures', 'bar')
    ]
  })
  expect(res).toMatchSnapshot()
})
