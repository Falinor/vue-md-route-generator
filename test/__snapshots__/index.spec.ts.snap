// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test 1`] = `
"function FooBarBazOverview() {
  return import(
    /* webpackChunkName: \\"foo-bar-baz-overview\\" */ '@/assets/foo/bar/baz/overview.md'
  )
}
function FooFooBar() {
  return import(/* webpackChunkName: \\"foo-foo-bar\\" */ '@/assets/foo/foo-bar.md')
}
function FooIndex() {
  return import(/* webpackChunkName: \\"foo-index\\" */ '@/assets/foo/index.md')
}

function BarIndex() {
  return import(/* webpackChunkName: \\"bar-index\\" */ '@/assets/bar/index.md')
}
function BarMeta() {
  return import(/* webpackChunkName: \\"bar-meta\\" */ '@/assets/bar/meta.md')
}

export default [
  {
    name: 'foo',
    path: '/foo',
    redirect: { name: 'foo-index' },
    children: [
      {
        name: 'foo-bar',
        path: '/foo/bar',
        redirect: { name: 'foo-bar-baz' },
        children: [
          {
            name: 'foo-bar-baz',
            path: '/foo/bar/baz',
            redirect: { name: 'foo-bar-baz-overview' },
            children: [
              {
                name: 'foo-bar-baz-overview',
                path: '/foo/bar/baz/overview',
                component: FooBarBazOverview
              }
            ]
          }
        ]
      },
      {
        name: 'foo-foo-bar',
        path: '/foo/foo-bar',
        component: FooFooBar
      },
      {
        name: 'foo-index',
        path: '/foo',
        component: FooIndex
      }
    ]
  },
  {
    name: 'bar',
    path: '/bar',
    redirect: { name: 'bar-index' },
    children: [
      {
        name: 'bar-index',
        path: '/bar',
        component: BarIndex
      },
      {
        name: 'bar-meta',
        path: '/bar/meta',
        component: BarMeta
      }
    ]
  }
]
"
`;
